File: tests/Domain/Subscription/SubscriptionTest.php
<?php

namespace Tests\Domain\Subscription;

use App\Domain\Identity\UserId;
use App\Domain\Subscriptions\Subscription;
use App\Domain\Subscriptions\SubscriptionId;
use App\Domain\Subscriptions\UserFollowed;
use Tests\Domain\FakeEventPublisher;

class SubscriptionTest extends \PHPUnit_Framework_TestCase
{
    public function testWhenFollowAnotherUser_ThenUserFollowedIsRaised()
    {
        $fakeEventPublisher = new FakeEventPublisher();
        $followeeId = new UserId('clem@mix-it.fr');
        $followerId = new UserId('jean@mix-it.fr');

        Subscription::followUser($fakeEventPublisher, $followerId, $followeeId);

        \Assert\that($fakeEventPublisher->events)->count(1);
        /** @var UserFollowed $userFollowed */
        $userFollowed = $fakeEventPublisher->events[0];
        \Assert\that($userFollowed->getSubscriptionId()->getFollowerId())->eq($followerId);
        \Assert\that($userFollowed->getSubscriptionId()->getFolloweeId())->eq($followeeId);
    }

    public function testWhenUnfollow_ThenUserUnfollowedIsRaised()
    {
        $fakeEventPublisher = new FakeEventPublisher();
        $followeeId = new UserId('clem@mix-it.fr');
        $followerId = new UserId('jean@mix-it.fr');
        $userFollowed = new UserFollowed(new SubscriptionId($followerId, $followeeId));
        $subscription = new Subscription(array($userFollowed));

        $subscription->unfollow($fakeEventPublisher);

        \Assert\that($fakeEventPublisher->events)->count(1);
        /** @var UserUnfollowed $userUnfollowed */
        $userUnfollowed = $fakeEventPublisher->events[0];
        \Assert\that($userUnfollowed)->isInstanceOf('App\Domain\Subscriptions\UserUnfollowed');
        \Assert\that($userUnfollowed->getSubscriptionId())->eq($userFollowed->getSubscriptionId());
    }
}

File: Mixter.Domain.Tests/Core/Subscriptions/SubscriptionTest.cs
using Mixter.Domain.Core.Messages;
using Mixter.Domain.Core.Subscriptions;
using Mixter.Domain.Core.Subscriptions.Events;
using Mixter.Domain.Identity;
using NFluent;
using Xunit;

namespace Mixter.Domain.Tests.Core.Subscriptions
{
    public class SubscriptionTest
    {
        private static readonly UserId Follower = new UserId("emilien@mixit.fr");
        private static readonly UserId Followee = new UserId("florent@mixit.fr");
        private static readonly SubscriptionId SubscriptionId = new SubscriptionId(Follower, Followee);

        private readonly EventPublisherFake _eventPublisher;

        public SubscriptionTest()
        {
            _eventPublisher = new EventPublisherFake();
        }

        [Fact]
        public void WhenFollowThenUserFollowedIsRaised()
        {
            Subscription.FollowUser(_eventPublisher, Follower, Followee);

            Check.That(_eventPublisher.Events).Contains(new UserFollowed(SubscriptionId));
        }

        [Fact]
        public void WhenUnfollowThenUserUnfollowedIsRaised()
        {
            var subscription = Create(new UserFollowed(SubscriptionId));

            subscription.Unfollow(_eventPublisher);

            Check.That(_eventPublisher.Events).Contains(new UserUnfollowed(SubscriptionId));
        }

        [Fact]
        public void WhenNotifyFollowerThenFolloweeMessageQuacked()
        {
            var subscription = Create(new UserFollowed(SubscriptionId));

            var messageId = MessageId.Generate();
            subscription.NotifyFollower(_eventPublisher, messageId);

            Check.That(_eventPublisher.Events).Contains(new FolloweeMessageQuacked(SubscriptionId, messageId));
        }

        [Fact]
        public void GivenUnfollowWhenNotifyFollowerThenDoNotRaisedFollowerMessageQuacked()
        {
            var subscription = Create(
                new UserFollowed(SubscriptionId),
                new UserUnfollowed(SubscriptionId));

            var messageId = MessageId.Generate();
            subscription.NotifyFollower(_eventPublisher, messageId);

            Check.That(_eventPublisher.Events).IsEmpty();
        }

        private Subscription Create(params IDomainEvent[] events)
        {
            return new Subscription(events);
        }
    }
}
